### ISSUE TITLE
Problem with image converter

### ISSUE THREAD

**Issue by IanSB** [simondotm/bbc-nula#1]
Hi Simon,
I don't know if you've been following the BBCtoHDMI thread on stardot but I've got a 16 colour mode 2 with programmable palette working with that and can display nula type mode 2 images with no internal hardware mods:
https://stardot.org.uk/forums/viewtopic.php?f=3&t=14430&start=600#p226594
However there is a problem with the converted images:
The BBCtoHDMI converter captures the BBC video output into a frame buffer on the Pi which is slightly larger than the displayed image because each mode is shifted differently in the horizontal direction and also to cope with *TV 255,0 etc. 
However the Pi can't tell the difference between logical colour 0 on the mode 2 display and the BBC black screen border so if logical colour 0 is programmed to anything other than black you get a coloured border around the image. You can see that in the BBC micro image linked above and I've also attached another example:
![car](https://user-images.githubusercontent.com/45300594/53825466-41899980-3f6e-11e9-812c-8cc32d793233.jpg)
Would you consider modding your converter and updating the images to always put the darkest colour on logical colour 0?

**Comment by simondotm** [comment]
Hi Ian,
Thanks for your note. Great work on the project btw.
It shouldn't be a problem to sort the palette before remapping so I'll take a look at that when I get a free moment. 
Out of interest, what will happen for example in an image such as this:
![image](https://user-images.githubusercontent.com/7404321/53891641-bb7b5a80-4022-11e9-8936-4957b1a4db44.png)
Where all 16-colours are used, but the image doesn't really have an obvious BG colour as such? Will the HDMI version have a pixel border based on whatever logical colour 0 is set to?
Intuitively an artist would assign the light blue colour to palette entry 0, but in an automated conversion there is a lack of subjectivity!

Best,
Simon

**Comment by simondotm** [comment]
Just to add to that, by sorting the palette, we'll get a black or dark border since the image contains dark palette entries, so I'm sure it'll will be an effective fix for most cases.
This one is probably a better example:
![image](https://user-images.githubusercontent.com/7404321/53892016-6be95e80-4023-11e9-89cf-286330c335bd.png)

As there's no 'dark' colours present. Mind you, the darkest colour is probably the blue BG .... but red *might* compete for that prize in which case we'd get a red border again. Hmmm..
Let's give it a go anyway.

**Comment by simondotm** [comment]
Perhaps I could also run a histogram to weight toward the most prevalent colour in the image. Interesting stuff.

**Comment by IanSB** [comment]
> As there's no 'dark' colours present. Mind you, the darkest colour is probably the blue BG .... but red might compete for that prize in which case we'd get a red border again. Hmmm..
Let's give it a go anyway.

Yes I noticed the problem with that one myself. Perhaps one option would be to look for the darkest colour and if there isn't a suitable colour reasonably close to black then scan the boundary of the image and use the most common colour on the boundary as logical 0 which should give blue in the above example.

Once I've got the command codes fully implemented I plan on modifying your beeb code to drive the palette in the converter so the demo should run nearly identically to a nula. I think it should be possible to auto switch between nula and hdmi mode by detecting the nula driver ROM

**Comment by simondotm** [comment]
Good idea to do the border scan. I'll incorporate that approach I think.

**Comment by IanSB** [comment]
Hi Simon,
I've been working quite a lot on the HDMI converter recently including adding a profile system and auto switching support for PC style resolutions so another suggestion I have for your image converter is to have a selectable output resolution. The HDMI converter can now display PC CGA and EGA modes (640x200 and 640x350 respectively) and they are essentially the same as bbc mode 0 which is being interpreted as mode 2 so I could run a similar demo on early PCs at 160x200 and 160x350. Both of those resoloutions would be scaled to 4:3 so the image would have to be squashed vertically for CGA and stretched vertically for EGA.

**Comment by IanSB** [comment]
I changed the way the frame buffer is implemented so it can now fill the entire active screen area and in many cases it looks quite good with a coloured border but the car one is still wrong I think:
![capture29](https://user-images.githubusercontent.com/45300594/54861352-9c1d4680-4d1f-11e9-8b11-cf9be295ff1b.png)

This one looks good with a border even though it has black:
![capture30](https://user-images.githubusercontent.com/45300594/54861363-afc8ad00-4d1f-11e9-8d27-b7625ac346da.png)

**Comment by IanSB** [comment]
Sorry, hit close by accident
